File: cpp/tests/test_gf_ladder.cpp
/**
 * GoldenFloat — Binary GF ladder C++ binding smoke tests (self-contained, no JSON dep).
 * MIT License — Copyright (c) 2026 Trinity Project
 */

#include <goldenfloat/gf_ladder.hpp>
#include <cassert>
#include <cmath>
#include <cstdio>

using namespace goldenfloat;

static bool approx(float a, float b, float tol) {
    return std::fabs(a - b) / (std::fabs(b) + 1e-9f) <= tol;
}

// Values inside every rung's normal range (Gf8 is tightest: ~[0.25, 15.5]).
template <class GF>
static void check_rung(float rt_tol, float op_tol) {
    for (float v : {0.5f, 1.0f, 1.5f, 2.0f, 3.0f, -2.5f, 4.0f}) {
        assert(approx(GF::from_f32(v).to_f32(), v, rt_tol));
    }
    auto a = GF::from_f32(1.5f);
    auto b = GF::from_f32(2.5f);
    assert(approx((a + b).to_f32(), 4.0f, op_tol));
    assert(approx((b - a).to_f32(), 1.0f, op_tol));
    assert(approx((a * b).to_f32(), 3.75f, op_tol));
    assert(approx((b / a).to_f32(), 2.5f / 1.5f, op_tol));
    assert(approx((-a).to_f32(), -1.5f, op_tol));
    assert(approx((-a).abs().to_f32(), 1.5f, op_tol));
    assert(GF::from_f32(1.0f).is_finite());
    assert(!GF::from_f32(INFINITY).is_finite()); // inf overflows on every rung
}

int main() {
    check_rung<Gf8>(0.05f, 0.05f);
    check_rung<Gf12>(0.01f, 0.01f);
    check_rung<Gf20>(0.001f, 0.001f);
    check_rung<Gf24>(0.0005f, 0.0005f);
    check_rung<Gf32>(0.0002f, 0.0002f);

    // carrier widths
    assert(Gf8::from_f32(1.5f).raw() <= 0xFFu);
    assert(Gf12::from_f32(1.5f).raw() <= 0xFFFu);

    std::printf("Binary GF ladder C++ bindings: ALL PASS\n");
    return 0;
}

File: python/goldenfloat/tests/test_gf_ladder.py
"""Binary GF ladder Python binding smoke tests (Gf8/Gf12/Gf20/Gf24/Gf32)."""
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))
from goldenfloat import Gf8, Gf12, Gf20, Gf24, Gf32

# (class, roundtrip tol, arithmetic tol) — tol scales with mantissa width.
RUNGS = [
    (Gf8, 0.05, 0.05),
    (Gf12, 0.01, 0.01),
    (Gf20, 0.001, 0.001),
    (Gf24, 0.0005, 0.0005),
    (Gf32, 0.0002, 0.0002),
]

# Values inside every rung's normal range (Gf8 is the tightest: ~[0.25, 15.5]).
VALUES = [0.5, 1.0, 1.5, 2.0, 3.0, -2.5, 4.0]


def test_roundtrip():
    for cls, rt_tol, _ in RUNGS:
        for v in VALUES:
            q = cls.from_f32(v).to_f32()
            assert abs(q - v) / (abs(v) + 1e-9) < rt_tol, (cls.__name__, v, q)


def test_arithmetic():
    for cls, _, tol in RUNGS:
        a = cls.from_f32(1.5)
        b = cls.from_f32(2.5)
        assert abs((a + b).to_f32() - 4.0) < tol, cls.__name__
        assert abs((b - a).to_f32() - 1.0) < tol, cls.__name__
        assert abs((a * b).to_f32() - 3.75) < tol, cls.__name__
        assert abs((b / a).to_f32() - (2.5 / 1.5)) < tol, cls.__name__
        assert abs((-a).to_f32() + 1.5) < tol, cls.__name__
        assert abs(abs(-a).to_f32() - 1.5) < tol, cls.__name__


def test_finiteness():
    for cls, _, _ in RUNGS:
        assert cls.from_f32(1.0).is_finite(), cls.__name__
        # inf overflows to the reserved Inf row on every rung.
        assert not cls.from_f32(float("inf")).is_finite(), cls.__name__


if __name__ == "__main__":
    test_roundtrip()
    test_arithmetic()
    test_finiteness()
    print("Binary GF ladder python bindings: ALL PASS")
